### ISSUE TITLE
Upload many files concurrently using Crt s3 java client: Exception thrown in S3MetaRequest#close, ignoring

### ISSUE THREAD

**Issue by ForgQi** [aws/aws-sdk-java-v2#3658]
### Describe the bug

I am using 256 threads for concurrent uploads in spring boot and some requests are getting error logs

### Expected Behavior

I don't know, maybe there is no error thrown

### Current Behavior

```
java.lang.NullPointerException: Cannot invoke "software.amazon.awssdk.crt.s3.S3MetaRequest.close()" because "this.metaRequest" is null
        at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.lambda$completeFutureAndCloseRequest$2(S3CrtResponseHandlerAdapter.java:113) ~[s3-2.19.4.jar!/:na]
        at software.amazon.awssdk.utils.FunctionalUtils.runAndLogError(FunctionalUtils.java:40) ~[utils-2.19.4.jar!/:na]
        at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.completeFutureAndCloseRequest(S3CrtResponseHandlerAdapter.java:112) ~[s3-2.19.4.jar!/:na]
        at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.lambda$onSuccessfulResponseComplete$1(S3CrtResponseHandlerAdapter.java:106) ~[s3-2.19.4.jar!/:na]
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[na:na]
        at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(Unknown Source) ~[na:na]
        at java.base/java.util.concurrent.CompletableFuture.whenComplete(Unknown Source) ~[na:na]
        at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.onSuccessfulResponseComplete(S3CrtResponseHandlerAdapter.java:100) ~[s3-2.19.4.jar!/:na]
        at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.onFinished(S3CrtResponseHandlerAdapter.java:95) ~[s3-2.19.4.jar!/:na]
        at software.amazon.awssdk.crt.s3.S3MetaRequestResponseHandlerNativeAdapter.onFinished(S3MetaRequestResponseHandlerNativeAdapter.java:24) ~[aws-crt-0.20.4.jar!/:0.20.4]
```

### Reproduction Steps

```java
  CompletableFuture<PutObjectResponse> future = client.putObject(objectRequest,
          AsyncRequestBody.fromFile(Paths.get(path))
  );
  future.whenComplete((resp, err) -> {
        if (resp != null) {
            System.out.println("Object uploaded. Details: " + resp);
        } else {
            // Handle error
            err.printStackTrace();
        }
  });

  future.join();
```

### AWS Java SDK version used

2.19.4

### JDK version used

openjdk version "17.0.4" 2022-07-19

### Operating System and version

Ubuntu 16.04.7 LTS

**Comment by zoewangg** [comment]
Hi @ForgQi, it seems it was logged from this line. https://github.com/aws/aws-sdk-java-v2/blob/master/services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtResponseHandlerAdapter.java#L112 
Did the requests actually fail?

**Comment by siva-subramani** [comment]
I am seeing this issue with the 2.0 version of AWS SDK, followed by these messages it failed by saying "Failed to send the request: Response code indicates throttling". any thoughts?

**Comment by charlesritchea** [comment]
Just saw this message with the latest aws java libs. No helpful context to indicate if the request succeeded or not

**Comment by emmanueldeviller** [comment]
I also got the same exception with aws-crt v0.29.9 when uploading files to S3 server. 

```
Cannot invoke "software.amazon.awssdk.crt.s3.S3MetaRequest.close()" because "this.metaRequest" is null
	at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.lambda$completeFutureAndCloseRequest$2(S3CrtResponseHandlerAdapter.java:125)
	at software.amazon.awssdk.utils.FunctionalUtils.runAndLogError(FunctionalUtils.java:40)
	at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.completeFutureAndCloseRequest(S3CrtResponseHandlerAdapter.java:124)
	at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.lambda$onSuccessfulResponseComplete$1(S3CrtResponseHandlerAdapter.java:118)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:887)
	at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2357)
	at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.onSuccessfulResponseComplete(S3CrtResponseHandlerAdapter.java:112)
	at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.onFinished(S3CrtResponseHandlerAdapter.java:107)
	at software.amazon.awssdk.crt.s3.S3MetaRequestResponseHandlerNativeAdapter.onFinished(S3MetaRequestResponseHandlerNativeAdapter.java:25)
```

**Comment by peterson-dc** [comment]
I am seeing this exact NPE when using the CRT S3 client with the S3TransferManager for copying files asynchronously to S3.

**Comment by emmanueldeviller** [comment]
We encountered a NullPointerException (NPE) due to a timing issue when receiving a very rapid response from S3. This typically happens during testing with small objects and a local S3 server. The problem arises because the native s3ClientMakeMetaRequest function invokes the responseHandler before the S3CrtAsyncHttpClient can properly set the s3MetaRequest using the method responseHandler.metaRequest(s3MetaRequest).

To circumvent the NPE and prevent associated native resource leaks, we implemented a somewhat inelegant solution within the S3CrtResponseHandlerAdapter.completeFutureAndCloseRequest() method:

```
// Ugly hack
private void completeFutureAndCloseRequest() {
        resultFuture.complete(null);

        // Wait for metaRequest to be not null
        if (metaRequest == null) {
            for (int i = 0; i < 100; i++) {
                try {
                    Thread.sleep(10);
                    if (metaRequest != null) break;
                } catch (InterruptedException e) {
                    throw new RuntimeException(e);
                }
            }
        }
        runAndLogError(log.logger(), "Exception thrown in S3MetaRequest#close, ignoring",
                () -> metaRequest.close());
}
```
This workaround addresses the issue by waiting for metaRequest to become non-null before proceeding to close it, thus preventing the NPE and resource leakage.

In my opinion, it would be prudent to revisit this issue and consider alternative solutions.

**Comment by Venorcis** [comment]
> In my opinion, it would be prudent to revisit this issue and consider alternative solutions.

Agreed, we're seeing the same issue in the same situation still

**Comment by trevorhreed** [comment]
Can this be reopened? I think this is a legitimate issue.

**Comment by debora-ito** [comment]
Folks, please open a new Github issue and provide all the info you can in the template - a repro code, a stacktrace with the error, and some context on the transaction rate/TPS are the most important.

But before creating a new issue, check if the error still happens with the latest version of the SDK, `2.25.20` as of today. The S3 CRT Client have been receiving constant updates, just want to make sure the issue is not fixed already.

**Comment by trevorhreed** [comment]
I tried updating to `2.25.20` and I'm getting the following error. You can see an example project [here](https://github.com/trevorhreed/aws-sdk-bug/tree/main).

```
Exception in thread "main" java.lang.NoSuchMethodError: 'software.amazon.awssdk.crt.s3.S3ClientOptions software.amazon.awssdk.crt.s3.S3ClientOptions.withMultipartUploadThreshold(long)'
	at software.amazon.awssdk.services.s3.internal.crt.S3CrtAsyncHttpClient.createS3ClientOption(S3CrtAsyncHttpClient.java:102)
	at software.amazon.awssdk.services.s3.internal.crt.S3CrtAsyncHttpClient.<init>(S3CrtAsyncHttpClient.java:73)
	at software.amazon.awssdk.services.s3.internal.crt.S3CrtAsyncHttpClient.<init>(S3CrtAsyncHttpClient.java:64)
	at software.amazon.awssdk.services.s3.internal.crt.S3CrtAsyncHttpClient$Builder.build(S3CrtAsyncHttpClient.java:264)
	at software.amazon.awssdk.services.s3.internal.crt.S3CrtAsyncHttpClient$Builder.buildWithDefaults(S3CrtAsyncHttpClient.java:270)
	at software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder.lambda$resolveAsyncHttpClient$15(SdkDefaultClientBuilder.java:413)
	at java.base/java.util.Optional.map(Optional.java:260)
	at software.amazon.awssdk.utils.Either.lambda$map$0(Either.java:51)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at software.amazon.awssdk.utils.Either.map(Either.java:51)
	at software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder.lambda$resolveAsyncHttpClient$16(SdkDefaultClientBuilder.java:413)
	at java.base/java.util.Optional.map(Optional.java:260)
	at software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder.resolveAsyncHttpClient(SdkDefaultClientBuilder.java:413)
	at software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder.lambda$finalizeAsyncConfiguration$3(SdkDefaultClientBuilder.java:296)
	at software.amazon.awssdk.utils.AttributeMap$DerivedValue.primeCache(AttributeMap.java:600)
	at software.amazon.awssdk.utils.AttributeMap$DerivedValue.get(AttributeMap.java:589)
	at software.amazon.awssdk.utils.AttributeMap$Builder.resolveValue(AttributeMap.java:396)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at software.amazon.awssdk.utils.AttributeMap$Builder.build(AttributeMap.java:362)
	at software.amazon.awssdk.core.client.config.SdkClientConfiguration$Builder.build(SdkClientConfiguration.java:232)
	at software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder.finalizeAsyncConfiguration(SdkDefaultClientBuilder.java:298)
	at software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder.asyncClientConfiguration(SdkDefaultClientBuilder.java:220)
	at software.amazon.awssdk.services.s3.DefaultS3AsyncClientBuilder.buildClient(DefaultS3AsyncClientBuilder.java:51)
	at software.amazon.awssdk.services.s3.DefaultS3AsyncClientBuilder.buildClient(DefaultS3AsyncClientBuilder.java:27)
	at software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder.build(SdkDefaultClientBuilder.java:155)
	at software.amazon.awssdk.services.s3.internal.crt.DefaultS3CrtAsyncClient.initializeS3AsyncClient(DefaultS3CrtAsyncClient.java:149)
	at software.amazon.awssdk.services.s3.internal.crt.DefaultS3CrtAsyncClient.<init>(DefaultS3CrtAsyncClient.java:83)
	at software.amazon.awssdk.services.s3.internal.crt.DefaultS3CrtAsyncClient.<init>(DefaultS3CrtAsyncClient.java:77)
	at software.amazon.awssdk.services.s3.internal.crt.DefaultS3CrtAsyncClient$DefaultS3CrtClientBuilder.build(DefaultS3CrtAsyncClient.java:313)
	at software.amazon.awssdk.services.s3.internal.crt.DefaultS3CrtAsyncClient$DefaultS3CrtClientBuilder.build(DefaultS3CrtAsyncClient.java:184)
	at AwsExample.main(AwsExample.java:15)

Process finished with exit code 1
```

**Comment by charlesritchea** [comment]
Update both AWS v2 ask _and_ AWS CRT

**Comment by bryanlb** [comment]
We're seeing this same issue as well, using SDK version `2.25.31`

**Comment by aymkhalil** [comment]
I've seen the same issue with SDK `2.25.40` and CRT `0.29.18` - I don't have a simple repro of it, but it happens with local minio server. Here is the stack trace I could capture:

```
Caused by: software.amazon.awssdk.core.exception.SdkClientException: Unexpected exception occurred: s3metaRequest is not initialized yet
	at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build(SdkClientException.java:111)
	at software.amazon.awssdk.core.exception.SdkClientException.create(SdkClientException.java:43)
	at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.lambda$onResponseBody$1(S3CrtResponseHandlerAdapter.java:114)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883)
	at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2251)
	at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.onResponseBody(S3CrtResponseHandlerAdapter.java:105)
	at software.amazon.awssdk.crt.s3.S3MetaRequestResponseHandlerNativeAdapter.onResponseBody(S3MetaRequestResponseHandlerNativeAdapter.java:19)]}
		at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:39)
		at org.junit.jupiter.api.Assertions.fail(Assertions.java:134)
		at com.datastax.cndb.test.BaseContainerTest.checkErrorLogs(BaseContainerTest.java:173)
		at com.datastax.cndb.test.BaseContainerTest$TestManager.afterEach(BaseContainerTest.java:271)
		at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeAfterEachCallbacks$12(TestMethodTestDescriptor.java:257)
		at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeAllAfterMethodsOrCallbacks$13(TestMethodTestDescriptor.java:273)
		at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
		at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeAllAfterMethodsOrCallbacks$14(TestMethodTestDescriptor.java:273)
		at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
		at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeAllAfterMethodsOrCallbacks(TestMethodTestDescriptor.java:272)
		at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeAfterEachCallbacks(TestMethodTestDescriptor.java:256)
		at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:141)
```

**Comment by RenatSh** [comment]
Seems to be fixed in 2.26.29: https://newreleases.io/project/github/aws/aws-sdk-java-v2/release/2.26.29
